synapseclient/cache.py

```diff
@@ -173,9 +174,27 @@ def get(self, file_handle_id, path=None):
             if path is not None:
                 ## If we're given a path to a directory, look for a cached file in that directory
                 if os.path.isdir(path):
-                    for cached_file_path, cached_time in six.iteritems(cache_map):
+                    matching_unmodified_directory = None
+                    removed_entry_from_cache = False # determines if cache_map needs to be rewritten to disk
+
+                    # iterate a copy of cache_map to allow modifying original cache_map
+                    for cached_file_path, cached_time in six.iteritems(dict(cache_map)):
                         if path == os.path.dirname(cached_file_path):
-                            return cached_file_path if compare_timestamps(_get_modified_time(cached_file_path), cached_time) else None
+                            if (os.path.exists(cached_file_path)
+                                and compare_timestamps(_get_modified_time(cached_file_path), cached_time)):
+                                    matching_unmodified_directory = cached_file_path
```

Comment:
The explicit check is redundant but should be faster (`compare_timestamps` returns `False` if either value is `None`)